### DISCUSSION TITLE
CODEOWNERS feature is dangerously broken

### DISCUSSION THREAD

**Discussion by piegamesde** [community/community#6927] [Pull Requests]
See https://github.com/NixOS/nixpkgs/issues/143441 for more information. In short:

- People are not getting pinged if they don't have write access to the repository. This includes organization teams. ~~The failure is silent.~~ that bit has been resolved, there is a lint for wrong entries when editing the file
- People are easily getting mass-pinged by accident
   - Unlike for example the new CI restrictions, new contributors are not exempt from this, raising DOS concerns.

To improve the situation:

- Allow people and teams without write access as CODEOWNERS
- Somehow notify when modifications to the file are syntactically incorrect instead of silently failing
- Add a protection against accidentally pinging too many people (e.g. manual approval by administrators).
  - Note that pinging everybody *may* be desired, it just shouldn't be that easy to *accidentally* do so.
- Require manual approval for first time contributors.

**Comment by piegamesde** [comment]
Point in case: the `nixpkgs` folks are not the only ones with the problem: https://about.sourcegraph.com/blog/a-different-way-to-think-about-code-ownership/ <- This excellent blog posts demonstrates the issues larger organizations face with the CODEOWNERS feature, and they even built a custom GitHub Action that re-implements it in a way that suits their needs.

**Reply by jakub-g** [reply]
> People are not getting pinged if they don't have write access to the repository.

We noticed that too. For example, you may have all people individually have write access to repo, but the `@team/something` doesn't have explicit push permission _as a team_ -> no CODEOWNERS notifs!

> People are easily getting mass-pinged by accident

Very easy when you open a PR towards wrong branch, you end up with a mega-PR with hundreds of commits and it pages whole company basically, dozens of teams. Potential fix: If there'd be say, more than 10 codeowners to be pinged on a PR, don't do it automatically, but display a prominent warning in UI on PR page: "This PR would notify 70 codeowners. Do you want to proceed"?

Also: currently if all those people are pinged, and then the issue is fixed (by changing destination branch of the PR for example), all those faulty codeowners stay in place, and there's no way to mass-remove them, apart one by one
